nixos/modules/services/networking/crab-hole.nix

```
@@ -0,0 +1,230 @@
+              "https://s3.amazonaws.com/lists.disconnect.me/simple_tracking.txt"
+              "file:///blocked.txt"
+            ];
+          };
+          upstream = {
+            name_servers = [
+              {
+                protocol = "tls";
+                socket_addr = "[2606:4700:4700::1111]:853";
+                tls_dns_name = "1dot1dot1dot1.cloudflare-dns.com";
+                trust_nx_responses = false;
+              }
+              {
+                protocol = "tls";
+                socket_addr = "1.1.1.1:853";
+                tls_dns_name = "1dot1dot1dot1.cloudflare-dns.com";
+                trust_nx_responses = false;
+              }
+            ];
+            options = {
+              validate = false;
+            };
+          };
+        };
+
+        type = lib.types.submodule {
+          freeformType = settingsFormat.type;
+          options = {
+            blocklist = {
+              include_subdomains = lib.mkEnableOption "Include Subdomains";
+              lists = lib.mkOption {
+                type = lib.types.listOf (lib.types.either lib.types.str lib.types.path);
+                default = [ ];
+                description = "List of blocklists. If files are added via url, make sure the service has access to them!";
+                apply = x: (lib.lists.forEach x (v: if builtins.isPath v then ("file://${v}") else v));
+              };
+              allow_list = lib.mkOption {
+                type = lib.types.listOf (lib.types.either lib.types.str lib.types.path);
+                default = [ ];
+                description = "List of allowlists. If files are added via url, make sure the service has access to them!";
+                apply = x: (lib.lists.forEach x (v: if builtins.isPath v then ("file://${v}") else v));
+              };
+            };
+
+            downstream = lib.mkOption {
+              description = "Configuration for the downstream DNS Servers";
+              type = lib.types.listOf (
+                lib.types.submodule {
+                  freeformType = settingsFormat.type;
+                  options = {
+                    protocol = lib.mkOption {
+                      type = lib.types.enum [
+                        "udp"
+                        "tls"
+                        "https"
+                        "quic"
+                      ];
+                      example = "udp";
+                      default = "udp";
+                      description = "The protocol to listen on";
+                    };
+                    listen = lib.mkOption {
+                      type = lib.types.str;
+                      example = "[::]";
+                      default = "localhost";
+                      description = "The IP to listen on";
+                    };
+                    port = lib.mkOption {
+                      type = lib.types.ints.u16;
```

We have a builtin type just for ports
```suggestion
                      type = lib.types.port;
```